Alerting: Support org level isolation of notification configuration

papagian opened the pull request:
<!--

Thank you for sending a pull request! Here are some tips:

1. If this is your first time, please read our contribution guide at https://github.com/grafana/grafana/blob/main/CONTRIBUTING.md

2. Ensure you include and run the appropriate tests as part of your Pull Request.

3. In a new feature or configuration option, an update to the documentation is necessary. Everything related to the documentation is under the docs folder in the root of the repository.

4. If the Pull Request is a work in progress, make use of GitHub's "Draft PR" feature and mark it as such.

5. If you can not merge your Pull Request due to a merge conflict, Rebase it. This gets it in sync with the main branch.

6. Name your PR as "<FeatureArea>: Describe your change", e.g. Alerting: Prevent race condition. If it's a fix or feature relevant for the changelog describe the user impact in the title. The PR title is used to auto-generate the changelog for issues marked with the "add to changelog" label.

-->

**What this PR does / why we need it**:
Database modifications for enabling organisation level isolation in Grafana 8 alerts.
It introduces a new migration that deletes all Grafana 8 alerting data and regenerates them so that on Alertmanager configuration is created properly for each organisation. **This has as side-effect that any rules or Alertmanager configuration created or modified manually after enabling the feature will be removed and recreated.** 

**Which issue(s) this PR fixes**:

<!--

- Automatically closes linked issue when the Pull Request is merged.

Usage: "Fixes #<issue number>", or "Fixes (paste link of issue)"

-->

Fixes #37263

**Special notes for your reviewer**:
With the present modifications the Alertmanager runs only for the main organisation (with ID = 1). Alertmanager support for all organisations will be addressed in follow up PRs.

# Release notice breaking change

### Unified Alerting (Grafana 8 Alerting) data loss 

Grafana v8.2 fixed an issue with org isolation for notification configuration but to fix this Grafana will now re-run the migration from old alerting and this will cause complete removal of all new alert rules and notification configurations. This data loss is not something we find acceptable and are working on ways to mitigate it. So in the meantime, if you are an early adopter of unified alerting please wait with trying v8.2 beta.

papagian committed:
Alerting: modify table and accessors to limit org access appropriately

papagian committed:
Update migration to create multiple Alertmanager configs

papagian committed:
Merge remote-tracking branch 'origin/main' into papagian/am-org-isolation-db

papagian committed:
fixup migration

papagian committed:
fixup: set OrgID everywhere

papagian committed:
Merge remote-tracking branch 'origin/main' into papagian/am-org-isolation-db

papagian committed:
fix failing test

With the current changes and until org isolation is properly fixed
the Alertmanager is running only for the main organisation (with ID 1).
Therefore the user making the request has to belong to this organisation.

papagian committed:
lint

papagian commented on pkg/services/sqlstore/migrations/ualert/tables.go:
```diff
@@ -264,4 +264,14 @@ func AddAlertmanagerConfigMigrations(mg *migrator.Migrator) {
+
+	// add org_id column; set default to 1 (Main org.)
+	mg.AddMigration("add column org_id in alert_configuration", migrator.NewAddColumnMigration(alertConfiguration, &migrator.Column{
+		Name: "org_id", Type: migrator.DB_BigInt, Nullable: false, Default: "0",
```
Setting default to zero has the implication that configurations created by the migration before this change will become invisible to the users (users have to re-run the migration by disabling-enabling the `ngalert` feature) to get their configurations back.

However I prefer this than setting the default to the 1 (=the main organisation ID).

Any suggestions for handling better this are welcome!

kylebrandt replied:
Alerting objects (contact points, policies) may have been created outside the dashboard->ngalert migration after the enabling of the feature toggle. The proper org for these would be unknown. Therefore all config must be cleared with the adding of Orgs for things to be in a valid state. If this is correct, then the main issue I think is: Does Grafana take care of this, or does a user have to do it as part of the upgrade? 

It seems more desirable to me for the user (or some subset) not have to do this. If it is not required that a user do it, then it seems like it should be an additional migration entry. However, since everything needs to be cleared anyways, I think it is maybe okay to modify the migration code and point to the same code with two entries. I believe this would effectively make it so the migration is done twice going forward though, which is perhaps not good.

papagian replied:
I agree that all Alertmanager configuration should be cleared (since their organisation is unknown) and I like the idea of doing it automatically without the users to take any extra action. But I think that we need to do it in a way that users will not loose also the rules they have created as well.

gotjosh replied:
I've left a comment for this on Slack, let me know if you want me to expand here.

papagian replied:
We have decided to force re-running the migration automatically without having the users to do anything.

papagian committed:
more linting

gotjosh commented on pkg/services/ngalert/notifier/alertmanager.go:
```diff
@@ -72,6 +73,8 @@ const (
 	}
 }
 `
+	// temporary until fix org isolation
```
```suggestion
	//TODO: temporary until fix org isolation
```

papagian replied:
I suggest better adding this as a part of #37320
In this one I have focused on the DB modifications and I have introduced only the necessary Alertmanager modification so that tests and the server run without failures for the main organisation.

gotjosh commented on pkg/services/sqlstore/migrations/ualert/ualert.go:
```diff
@@ -343,8 +336,14 @@ func (m *rmMigration) Exec(sess *xorm.Session, mg *migrator.Migrator) error {
 
-	if err := os.RemoveAll(silencesFileName(mg)); err != nil {
-		mg.Logger.Error("alert migration error: failed to remove silence file", "err", err)
+	files, err := getAllSilenceFiles(mg)
```
Remind me again why do we remove the silences file?

papagian replied:
This task runs during [the backward migration](https://github.com/grafana/grafana/blob/90572dc7219ae3fdaeaa46f6a133c1e9b60479b5/pkg/services/sqlstore/migrations/ualert/ualert.go#L68) (when disabling `ngalert`) so we clean everything related to the new alerting.

gotjosh commented on pkg/tests/api/alerting/api_alertmanager_configuration_test.go:
```diff
@@ -71,12 +76,15 @@ func TestAlertmanagerConfigurationIsTransactional(t *testing.T) {
 	})
 
 	store := testinfra.SetUpDatabase(t, dir)
+	// override bus to get the GetSignedInUserQuery handler
```
I'd love to see a test where we have two different configurations for two users from different organisations.

papagian replied:
I have added the test (together with some fixes/improvements)!

gotjosh reviewed (commented):
The approach looks solid, will give it another round once my comments are addressed.

papagian commented on pkg/services/sqlstore/migrations/ualert/channel.go:
```diff
@@ -50,25 +58,31 @@ func (m *migration) getNotificationChannelMap() (map[interface{}]*notificationCh
+			if _, ok := defaultChannelsMap[c.OrgID]; !ok { // new seen org
+				defaultChannelsMap[c.OrgID] = make([]*notificationChannel, 0, len(allChannels))
+			}
 			// TODO: verify that there will be only 1 default channel.
```
I don't see any reason to limit the default channels to only one. In my local setup I have two and in the legacy alerting we supported that.
So I suggest removing the TODO.

codesome replied:
Yes the TODO can be removed. I had assumed we can have only 1 default channel before, and I was wrong :)

papagian committed:
Apply suggestions from code review

Co-authored-by: gotjosh <josue@grafana.com>

papagian committed:
fix renamed variables

papagian committed:
Add test and small fix

papagian commented on pkg/services/ngalert/notifier/alertmanager.go:
```diff
@@ -399,7 +405,7 @@ func (am *Alertmanager) applyConfig(cfg *apimodels.PostableUserConfig, rawConfig
 
 func (am *Alertmanager) WorkingDirPath() string {
-	return filepath.Join(am.Settings.DataPath, workingDir)
+	return filepath.Join(am.Settings.DataPath, workingDir, strconv.Itoa(mainOrgID))
```
The path under which the templates and silences will be stored will eventually be:
[paths.data configuration](https://grafana.com/docs/grafana/latest/administration/configuration/?plcmt=footer#data)/alerting/<org_id>
eg: `data/alerting/1`.
Is that what we want?

gotjosh replied:
That seems correct to me.

papagian committed:
Remove TODO

papagian committed:
Merge remote-tracking branch 'origin/main' into papagian/am-org-isolation-db

codesome commented on pkg/services/sqlstore/migrations/ualert/channel.go:
```diff
@@ -50,25 +58,30 @@ func (m *migration) getNotificationChannelMap() (map[interface{}]*notificationCh
 		if c.IsDefault {
-			// TODO: verify that there will be only 1 default channel.
-			defaultChannels = append(defaultChannels, &allChannels[i])
+			if _, ok := defaultChannelsMap[c.OrgID]; !ok { // new seen org
+				defaultChannelsMap[c.OrgID] = make([]*notificationChannel, 0, len(allChannels))
+			}
```
I think this is not required. The below line should be enough, it wont cause any panics. (and `len(allChannels)` will be bit excessive although it wont matter)

codesome commented on pkg/services/sqlstore/migrations/ualert/channel.go:
```diff
@@ -220,32 +238,40 @@ func makeKeyForChannelGroup(channelUids map[interface{}]struct{}) (string, error
-		return err
-	}
+func (m *migration) addDefaultChannels(amConfigsPerOrg amConfigsPerOrg, allChannels channelsPerOrg, defaultChannels defaultChannelsPerOrg) error {
+	for orgID := range allChannels {
```
If there is some org without any notification channels yet, then no default route will be added to that org here. But I think we need at least 1 default routing policy (and a contact point to attach to that) per org. If we have a list of all orgs, then it would be easiest to iterate through that list here instead of `allChannels` (which will end up adding an empty contact point and default routing policy for such orgs).

papagian replied:
How the result configuration should look like in case there are not any notification channels associated with an organisation?
Does the following makes sense?
```
{
	"template_files": null,
	"alertmanager_config": {
		"route": {
			"receiver": "autogen-contact-point-default"
		},
		"templates": null,
		"receivers": [
			{
				"name": "autogen-contact-point-default"
			}
		]
	}
}
```

codesome replied:
You are right, it is best to store nothing in that case and go with the default config that we store. Please ignore this comment :)

codesome commented on pkg/services/sqlstore/migrations/ualert/channel.go:
```diff
@@ -220,32 +238,40 @@ func makeKeyForChannelGroup(channelUids map[interface{}]struct{}) (string, error
 }
 
-func (m *migration) addUnmigratedChannels(amConfig *PostableUserConfig, allChannels map[interface{}]*notificationChannel, defaultChannels []*notificationChannel) error {
+func (m *migration) addUnmigratedChannels(orgID int64, amConfigsPerOrg *PostableUserConfig, allChannels map[interface{}]*notificationChannel, defaultChannels []*notificationChannel) error {
```
```suggestion
func (m *migration) addUnmigratedChannels(orgID int64, amConfig *PostableUserConfig, allChannels map[interface{}]*notificationChannel, defaultChannels []*notificationChannel) error {
```

codesome reviewed (commented):
LGTM on the notification channel migration. Just a couple of nits and a question/fix.

papagian committed:
Apply suggestions from code review

papagian committed:
Apply suggestions from code review

papagian committed:
Force re-running the dashboard alert migration

gotjosh reviewed (approved):
LGTM.

kylebrandt commented:
On my initial run (feature flag enabled, and was enabled), I get:

`EROR[08-11|09:38:10] Executing migration failed               logger=migrator id="clone move dashboard alerts to unified alerting" error="no configuration found for org: 1"`

looking to see if I can figure out why

papagian commented:
> On my initial run (feature flag enabled, and was enabled), I get:
> 
> `EROR[08-11|09:38:10] Executing migration failed logger=migrator id="clone move dashboard alerts to unified alerting" error="no configuration found for org: 1"`
> 
> looking to see if I can figure out why

do you have notification channels in org 1?  strange also why the "move dashboard alerts to unified alerting" does not fail with the same error (or maybe this migration had already run before switching to this branch?)

kylebrandt commented:
> do you have notification channels in org 1? 

I do not. It seems the db I copied to my laptop was one that had some rules in the `alert` tables (old alerting), but none of them have any notifications assigned. These were the rules that come from gdev dashboards.

> strange also why the "move dashboard alerts to unified alerting" does not fail with the same error (or maybe this migration had already run before switching to this branch?)

The migration has already run before switching to the branch.

papagian committed:
Fix migration when there are no notification channels

papagian committed:
replace mg.ClearMigrationEntry()

mg.ClearMigrationEntry() would create a new session.
This commit introduces a new migration for clearing an entry from migration log for replacing  mg.ClearMigrationEntry() so that all dashboard alert migration operations will run inside the same transaction.
It adds also `SkipMigrationLog()` in Migrator interface for skipping adding an entry in the migration_log.

papagian commented on pkg/services/sqlstore/migrations/ualert/tables.go:
```diff
@@ -264,4 +264,14 @@ func AddAlertmanagerConfigMigrations(mg *migrator.Migrator) {
 	mg.AddMigration("alert alert_configuration alertmanager_configuration column from TEXT to MEDIUMTEXT if mysql", migrator.NewRawSQLMigration("").
 		Mysql("ALTER TABLE alert_configuration MODIFY alertmanager_configuration MEDIUMTEXT;"))
+
+	// add org_id column; set default to 1 (Main org.)
```
good catch! I will remove it

papagian committed:
Remove obsolete comment

papagian committed:
Add comments

papagian commented on pkg/services/sqlstore/migrations/ualert/ualert.go:
```diff
@@ -49,37 +51,105 @@ func AddDashAlertMigration(mg *migrator.Migrator) {
+	switch {
+	case ngEnabled && !migrationRun:
+		// removes all unified alerting data
+		mg.AddMigration(cloneRmMigTitle, &rmMigrationWithoutLogging{})
```
This used to be:
```
                 mg.AddMigration(cloneRmMigTitle, &rmMigration{})

		// Remove the migration entry that removes all unified alerting data. This is so when the feature
		// flag is removed in future the "remove unified alerting data" migration will be run again.
		err = mg.ClearMigrationEntry(cloneRmMigTitle)
		if err != nil {
			mg.Logger.Error("alert migration error: could not clear clone alert migration for removing data", "error", err)
		}
```
but `mg.ClearMigrationEntry()` didn't have any effect (the entry didn't get deleted) because it starts a new session and the `rmMigration` runs in a transaction that is not committed yet,
so the modification was invisible to it.

papagian committed:
more comments

gotjosh commented on pkg/services/sqlstore/migrations/ualert/ualert.go:
```diff
@@ -288,6 +352,7 @@ func (m *migration) writeAlertmanagerConfig(amConfig *PostableUserConfig, allCha
 		// Since we are migration for a snapshot of the code, it is always going to migrate to
```
To answer the above TODO, we shouldn't because the config will be applied on service startup and before it is ready. My understanding is that migrations are run before we start up services.

papagian replied:
I don't fully understand this TODO but I confirm that the migrations run first (the `sqlstore` service that starts the migrations runs with [higher](https://github.com/grafana/grafana/blob/11d03f7961db3cd7dd613d94a843d1a40f270b19/pkg/services/sqlstore/sqlstore.go#L71) priority)

papagian committed:
Fix issue with not initialised routes

papagian committed:
Fix missing receiver from route

gotjosh reviewed (approved):
LGTM.

We have found and solved a few issues after offline discussions. Tested migrating back and forth multiple times while having multiple organisations with rules and notification channels.

It all works as expected. 👌
